### PR TITLE
Added uniform random walk on graphs

### LINKED ISSUE #65: Add algorithm for DeepWalk
#### Description of the problem

Once we are done adding basic framework for Graphs, we can start thinking of adding DeepWalk algorithm. 
We first need to decide the feasibility of implementation of the above algorithm. 
#### References/Other comments
[1] https://github.com/phanein/deepwalk

### PR DESCRIPTION
**Pull request by Saptashrungi**
#### References to other Issues or PRs or Relevant literature

"Fixes #65"

#### Brief description of what is fixed or changed
The test function is not yet complete entirely. I am not able to run tests locally as the coverage reports are giving me some errors. The assert statement in test function for random walk should give error but its not happening.

### CODE REVIEW

File: pydatastructs/graphs/algorithms.py [hunk 1]
```diff
@@ -548,3 +548,32 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+
+def random_walk(G, start, path_length, alpha = None):
+    import random
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:554]: Shift to top

File: pydatastructs/graphs/algorithms.py [hunk 2]
```diff
@@ -548,3 +548,32 @@ def strongly_connected_components(graph, algorithm):
...
+    random.seed(44)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:555]: No need to fix the seed here.

File: pydatastructs/graphs/algorithms.py [hunk 3]
```diff
@@ -548,3 +548,32 @@ def strongly_connected_components(graph, algorithm):
...
+    if start:
+        path = [start]
+    else:
+        path = [random.choices(G.vertices)]
+  
+
+    for i in range(0, path_length):
+        cur = path[-1]
+        degree = len(G.neighbors(cur))
+        probablity_list = degree*[1/degree]
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:566]: `int`*`list`?

File: pydatastructs/graphs/algorithms.py [hunk 4]
```diff
@@ -548,3 +548,32 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+
+def random_walk(G, start, path_length, alpha = None):
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:553]:
```suggestion
def random_walk(graphs, start, path_length, alpha = None):
```

**Comment by czgdp1807** [comment]
Please add docstrings as well.

**Comment by czgdp1807** [comment]
Shift `random_walk` to global namespace.

File: pydatastructs/graphs/algorithms.py (hunk 3 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:566]: Yes. If the list is of length equal to degree and the all elements in it are 1/degree.

**Comment by Saptashrungi** [comment]
I am not able to resolve this new error called AttributeError.

File: pydatastructs/graphs/tests/test_algorithms.py [hunk 5]
```diff
@@ -185,3 +185,42 @@ def _test_strongly_connected_components(func, ds, algorithm, *args):
...
+            (V4.name, V7.name),
+            (V4.name, V8.name)
+        ]
+
+        for edge in edges:
+            G1.add_edge(*edge)
+
+        random_walk = random_walk(G, V1, 3)
+
+        assert random_walk == []
```
Comment (czgdp1807) [inline pydatastructs/graphs/tests/test_algorithms.py:223]: Add a test where random walk isn't empty.

File: pydatastructs/graphs/algorithms.py [hunk 6]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+    else:
+        path = [random.choices(graphs.vertices)]
+
+
+    for i in range(0, path_length):
+        cur = path[-1]
+        degree = len(graphs.neighbors(cur))
+        probablity_list = degree*[1/degree]
+
+        if random.random() < alpha and alpha is not None:    #alpha determines (directly proportional) randomness of random walk
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:608]:
```suggestion
        if random.random() < alpha and alpha is not None:
```
References?

File: pydatastructs/graphs/algorithms.py [hunk 7]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> random_walk = random_walk(g, v1, 3)
+    >>> random_walk = ['v1', 'v2', 'v5']
+    True
+
+    """
+
+    if start:
+        path = [start]
+    else:
+        path = [random.choices(graphs.vertices)]
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:600]:
```suggestion
        path = [random.choices(graph.vertices)]
```

File: pydatastructs/graphs/algorithms.py [hunk 8]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> v1, v2, v3, v4, v5  = [AdjacencyListGraphNode(i) for i in range(5)]
+    >>> g = Graph(v1, v2, v3)
+    >>> g.add_edge(v1.name, v2.name)
+    >>> g.add_edge(v2.name, v3.name)
+    >>> g.add_edge(v3.name, v1.name)
+    >>> g.add_edge(v2.name, v4.name)
+    >>> g.add_edge(v4.name, v5.name)
+
+    >>> random_walk = random_walk(g, v1, 3)
+    >>> random_walk = ['v1', 'v2', 'v5']
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:592]:
```suggestion
    >>> random_walk == ['v1', 'v2', 'v5']
```

File: pydatastructs/graphs/algorithms.py [hunk 9]
```diff
@@ -126,7 +128,7 @@ def breadth_first_search_parallel(
         The rest of the arguments are optional and you can
         provide your own stuff there.
 
-    Note
+   Note
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:131]:
```suggestion
    Note
```

File: pydatastructs/graphs/algorithms.py [hunk 10]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+    Parameters
+    ==========
+
+    graph: Graph
+        The graph on which random walk needs to implemented.
+    start: Starting GraphNode
+        The node/vertex from where random walk will begin.
+    path_length:
+        The path length of random walk.
+    alpha:
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:568]: How does this parameter makes random walk more random? Any references?

File: pydatastructs/graphs/algorithms.py [hunk 11]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+def random_walk(graph, start, path_length, alpha = None):
+    """
+    Executes random walk from the starting node and given graph.
+
+    Parameters
+    ==========
+
+    graph: Graph
+        The graph on which random walk needs to implemented.
+    start: Starting GraphNode
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:564]:
```suggestion
    start: GraphNode or it's subclasses
```

File: pydatastructs/graphs/algorithms.py [hunk 12]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+
+def random_walk(graph, start, path_length, alpha = None):
+    """
+    Executes random walk from the starting node and given graph.
+
+    Parameters
+    ==========
+
+    graph: Graph
+        The graph on which random walk needs to implemented.
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:563]:
```suggestion
        The graph on which random walk needs to be executed.
```

File: pydatastructs/graphs/algorithms.py [hunk 13]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+    start: Starting GraphNode
+        The node/vertex from where random walk will begin.
+    path_length:
+        The path length of random walk.
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:567]:
```suggestion
        The maximum number of nodes/vertices traversed by random walk.
```

File: pydatastructs/graphs/algorithms.py [hunk 14]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+    alpha:
+        A parameter which determines the randomness. More the value, more random the path will 
+        be and less chances of restart.
+
+    Returns
+    =======
+
+    components: list
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:575]:
```suggestion
    path: list
```
`components` are for `minimum_spanning_tree`.

File: pydatastructs/graphs/algorithms.py [hunk 15]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
...
+        Python list of nodes after execution of random walk.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import Graph, AdjacencyListGraphNode
+    >>> from pydatastructs import strongly_connected_components
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:582]:
```suggestion
    >>> from pydatastructs import random_walk
```

File: pydatastructs/graphs/algorithms.py [hunk 16]
```diff
@@ -548,3 +550,68 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+
+def random_walk(graph, start, path_length, alpha = None):
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:555]:
```suggestion
def random_walk(graph, start, path_length, seed = 0, alpha = None):
```

File: pydatastructs/graphs/algorithms.py (hunk 8 above)
Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:592]: These tests will be non-deterministic. A change is need. See https://github.com/codezonediitj/pydatastructs/pull/246/files#r403713512

File: pydatastructs/graphs/algorithms.py (hunk 10 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:568]: Here `alpha` is a floating point number between 0 and 1. The `random.random()` function generates random floating number between 0 and 1. For example, if value if `alpha` is as low as 0.001, the chances are more that `random.random()` will generate a floating point number greater than `alpha`. As a result of this, the `RandomWalk` will not restart. On the other side, if `alpha` has value close to 1, the `RandomWalk`, will restart more frequently due to which it can go much away from the `start` node, reducing randomness.
This method to control randomness was used in random walk [here](https://github.com/phanein/deepwalk).

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:568]: Why would a random walk need a restart?

File: pydatastructs/graphs/algorithms.py [hunk 17]
```diff
@@ -188,7 +190,7 @@ def _breadth_first_search_parallel_adjacency_list(
                 status = Executor.submit(
                          _generate_layer, args=
                          (graph, node, layers[layer+1], visited,
-                          operation, *args), kwargs=kwargs).result()
+                          operation, args), kwargs=kwargs).result()
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:193]:
```suggestion
                          operation, *args), kwargs=kwargs).result()
```
Please avoid irrelevant changes.

File: pydatastructs/trees/heaps.py [hunk 18]
```diff
@@ -89,7 +89,7 @@ def __new__(cls, elements=None, heap_property="min", d=4):
         elif _check_type(elements, Array):
             elements = DynamicOneDimensionalArray(TreeNode, len(elements), elements._data)
         else:
-            raise ValueError(f'Expected a list/tuple/Array of TreeNode got {type(elements)}')
+            raise ValueError('Expected a list/tuple/Array of TreeNode got {type(elements)}')
```
Comment (czgdp1807) [inline pydatastructs/trees/heaps.py:92]:
```suggestion
            raise ValueError(f'Expected a list/tuple/Array of TreeNode got {type(elements)}')

```

File: pydatastructs/graphs/algorithms.py (hunk 17 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:193]: Yes, I was having pytest of older version installed, hence it gave error in these lines. Now the problem is rectified, I will modify it back.

**Comment by Saptashrungi** [comment]
I am not able to get solution for two errors I am facing. I use `graph.neighbors()` just like its used in the rest of the code, but I am still getting an error. `cur` is a string object I verified.

File: pydatastructs/graphs/algorithms.py [hunk 19]
```diff
@@ -548,3 +550,64 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> random_walk == ['v1', 'v2', 'v5']
+    True
+
+    """
+
+    path = [start.name]
+
+    for i in range(0, path_length):
+        cur = path[-1]
+        degree = len(graph.neighbors(cur))
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:600]:
```suggestion
```

File: pydatastructs/graphs/algorithms.py [hunk 20]
```diff
@@ -548,3 +550,64 @@ def strongly_connected_components(graph, algorithm):
...
+
+        if alpha is not None:
+            if random.random() < alpha:
+                next_node = start.name
+        else:
+            if graph.neighbors(cur) == []:
+                break
+            next_node = random.choices(graph.neighbors(cur))
+
+        path.append(next_node)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:611]:
```suggestion
        path.append(next_node.name)
```

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:611]: Though the API for neighbors is a bit inconsistent(names are not returned, a list of node objects is returned). IDK, why I did this. Now making a change by returning a list of names will require quite a good amount of effort. Let's leave it as it is.

File: pydatastructs/graphs/algorithms.py (hunk 10 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:568]: It is optional. If it has no use, I will remove this part.

File: pydatastructs/graphs/algorithms.py [hunk 21]
```diff
@@ -548,3 +550,64 @@ def strongly_connected_components(graph, algorithm):
...
+    True
+
+    """
+
+    path = [start.name]
+
+    for i in range(0, path_length):
+        cur = path[-1]
+        degree = len(graph.neighbors(cur))
+        probablity_list = []
```
[the remark this answers is not in the document]
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:601]: I modified it while debugging, forgot to reverse the changes.

File: pydatastructs/graphs/algorithms.py (hunk 19 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:600]: Everywhere else in the code the usage is the same. Why the error is here?

File: pydatastructs/graphs/algorithms.py [hunk 22]
```diff
@@ -548,3 +550,64 @@ def strongly_connected_components(graph, algorithm):
...
+
+        if alpha is not None:
+            if random.random() < alpha:
+                next_node = start.name
+        else:
+            if graph.neighbors(cur) == []:
+                break
+            next_node = random.choices(graph.neighbors(cur))
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:609]: I am not sure, whether `random.choices` is truly uniform. A simple empirical analysis will help figuring that thing out.

File: pydatastructs/graphs/algorithms.py (hunk 20 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:611]: `graph.neighbors()` returns list of strings right? Then `next_node` will be a string.

File: pydatastructs/graphs/algorithms.py (hunk 22 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:609]: There is one more parameter of weights, which is probability list, I removed it while debugging, I will put it back.

File: pydatastructs/graphs/algorithms.py (hunk 20 above)
Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:611]: > (names are not returned, a list of node objects is returned)

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:611]: The above suggestion may fix the error.

File: pydatastructs/graphs/algorithms.py (hunk 22 above)
Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py]: If `random.choices` picks uniformly from the list, then may be we do not need `probability_list`. 


> A simple empirical analysis will help figuring that thing out.

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py]:
```python
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> l = ["hey", "wassup", "byeeee", "timesupppppppppppp"]
>>> import random
>>> sample = [random.choices(l) for _ in range(1000)]
>>> sample.count(['hey'])
234
>>> sample.count(['wassup'])
257
>>> sample.count(['byeeee'])
236
>>> sample.count(['timesupppppppppppp'])
273
>>> sample = [random.choices(l) for _ in range(100000)]
>>> sample.count(['hey'])
24903
>>> sample.count(['wassup'])
24945
>>> sample.count(['byeeee'])
25246
>>> sample.count(['timesupppppppppppp'])
24906
>>> sample = [random.choices(l) for _ in range(1000000)]
>>> sample.count(['hey'])
249375
>>> sample.count(['wassup'])
250203
>>> sample.count(['byeeee'])
250014
>>> sample.count(['timesupppppppppppp'])
250408
>>> l = [1, 2, 3, -3, -2, -1]
>>> sample = [random.choices(l) for _ in range(1000000)]
>>> [sample.count([i]) for i in l]
[166668, 167257, 166685, 166324, 166541, 166525]
>>> 1000000/6
166666.66666666666
```
The above shows that, `random.choices` uniformly selects if `wieghts` is `None`. So no need to waste CPU cycles on computing `probability_list` or `degree`.

**Comment by Saptashrungi** [comment]
ping  @czgdp1807 
Why some tests are still pending? Is there some problem with the code or it takes time?

**Comment by czgdp1807** [comment]
May be pushing too many commits in a short span of time is causing this problem. It's a problem on Travis CI end.

**Comment by Saptashrungi** [comment]
Should I make a different PR?

**Comment by czgdp1807** [comment]
No, build restarted. Let's see if the status is updated.

P.S. I have some changes in my mind, will suggest sometime.

File: pydatastructs/graphs/algorithms.py [hunk 23]
```diff
@@ -548,3 +550,61 @@ def strongly_connected_components(graph, algorithm):
...
+    for i in range(0, path_length):
+        cur = path[-1]
+
+        if graph.neighbors(cur) == []:
+            break
+        next_node = random.choices(graph.neighbors(cur))
+
+        path.append(next_node[0].name)
+
+    return path
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:610]:
```suggestion
    """
    random.seed(seed)
    path = [start.name]

    for i in range(0, path_length):
        cur = path[-1]
        if graph.neighbors(cur) == []:
            break
        next_node = random.choices(graph.neighbors(cur))
        path.append(next_node[0].name)

    return path
```

File: pydatastructs/graphs/tests/test_algorithms.py [hunk 24]
```diff
@@ -185,3 +185,42 @@ def _test_strongly_connected_components(func, ds, algorithm, *args):
...
+            (V3.name, V6.name),
+            (V3.name, V7.name),
+            (V4.name, V7.name),
+            (V4.name, V8.name)
+        ]
+
+        for edge in edges:
+            G.add_edge(*edge)
+
+        random_walk_ = random_walk(G, V1, 3, seed = 44)
```
Comment (czgdp1807) [inline pydatastructs/graphs/tests/test_algorithms.py:221]:
```suggestion
        random_walk_ = random_walk(G, V1, 3, seed=44)
```

File: pydatastructs/graphs/algorithms.py [hunk 25]
```diff
@@ -548,3 +550,57 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> random_walk = random_walk(g, v1, 3, 44)
+    >>> random_walk == ['0', '2', '6']
+    True
+    """
+    random.seed(seed)
+    path = [start.name]
+
+    for i in range(0, path_length):
+        cur = path[-1]
+        if graph.neighbors(cur) == []:
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:601]:
```suggestion
        if not graph.neighbors(cur):
```

File: pydatastructs/graphs/algorithms.py [hunk 26]
```diff
@@ -548,3 +550,57 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> g.add_edge(v2.name, v5.name)
+    >>> g.add_edge(v2.name, v6.name)
+    >>> g.add_edge(v3.name, v6.name)
+    >>> g.add_edge(v3.name, v7.name)
+    >>> g.add_edge(v4.name, v7.name)
+    >>> g.add_edge(v4.name, v8.name)
+
+    >>> random_walk = random_walk(g, v1, 3, 44)
+    >>> random_walk == ['0', '2', '6']
+    True
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:594]: Use some simple example. A three node graph(v1->v2->v3->v1) would be better. It should say that nodes can be repeated in a random walk.

File: pydatastructs/graphs/algorithms.py [hunk 27]
```diff
@@ -548,3 +550,57 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+
+def random_walk(graph, start, path_length, seed = 0):
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:555]:
```suggestion
def random_walk(graph, start, path_length, seed=0):
```
No spaces around `=` in arguments of function/method.
In addition, I think doing a random walk is of no use unless we give user an opportunity to perform some operation while the random walk is at a given node. Currently, `breadth_first_search`(see the code in this file) does something like that. May be allowing an argument like operation will also help in implementing loop erased random walks by just defining `operation`.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:555]: In `breadth_first_search` we have `operation` as a function which is executed for each graph node. While in `random_walk` we can have `operation` of type string where it can take values `uniform` or `loop_erased`. User will give these values and `uniform` will be the default value.
Accordingly the corresponding function will be executed. Will this work ?

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:555]: I don't think that would make the code robust. `operation` should be a function.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:555]: If operation is a function, I will have to define the random walk function in the test function. Will it be okay?

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:555]: Well, few lines of code are always better than 1000 words of explanation. Please push the commits for showing what you are trying to do.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py]: ping @czgdp1807 
I just pushed a commit. For `loop_erased_random_walk` we just need to redefine `operation`.

File: pydatastructs/graphs/algorithms.py [hunk 28]
```diff
@@ -548,3 +549,45 @@ def strongly_connected_components(graph, algorithm):
...
+    >>> g.add_edge(v2.name, v3.name)
+    >>> g.add_edge(v3.name, v1.name)
+
+    >>> random_walk = random_walk(g, v1, 5, )
+    >>> random_walk == ['0', '1', '2', '0', '1']
+    True
+
+    '''
+
+    path = operation(graph, start, path_length)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:591]: This was not what I was talking about. Please see the implementation of `breadth_first_search`. `operation` should take, `curr_node`, `prev_node` and more stuff as it has been done in BFS.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:591]: In BFS, there is parent-child relation between current node and the next node, where `parent[next_node] == curr_node`. In `random walk`, if we have `operation`, we don't need the `next_node`( as we choose it randomly from neighbours), and the previous node can be found by `path[-2]`,(` path[-1]` being `curr_node`). I am not getting the analogy.

Comment (czgdp1807) [reply pydatastructs/graphs/algorithms.py:591]: Well, AFAICT, current node is just where you are. Tracing previous node is simple. Now, if you previous and current node are same then you found a loop. Moreover, we will not form a path, that can be done through the operation we are passing. operation will append nodes to the current path. That's just two examples of how we can use operation in random walks. So just the method of choosing next node is different. I hope you are getting my point.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py]: I pushed a commit which implements `loop_erased_random_walk`.
Also I have verified that `random.shuffle()` follows approximately uniform distribution as it follows
`fischer yates shuffle`.

File: pydatastructs/graphs/tests/test_algorithms.py [hunk 29]
```diff
@@ -185,3 +186,53 @@ def _test_strongly_connected_components(func, ds, algorithm, *args):
...
+        def walk(next_node, path):
+            path.append(next_node)
+            return True
+
+        def loop_erase(next_node, path):
+            if next_node in path:
+                return False
+            else:
+                path.append(next_node)
+                return True
```
Comment (czgdp1807) [inline pydatastructs/graphs/tests/test_algorithms.py:231]: Something simple. `loop_erase` can be added after work on this is complete.

File: pydatastructs/graphs/algorithms.py [hunk 30]
```diff
@@ -548,3 +550,61 @@ def strongly_connected_components(graph, algorithm):
         "isn't implemented for finding strongly connected components."
         %(algorithm, graph._impl))
     return getattr(algorithms, func)(graph)
+
+
+def random_walk(graph, start, path_length, operation, seed=44):
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:555]: How can I pass the parameters for operation?

File: pydatastructs/graphs/algorithms.py [hunk 31]
```diff
@@ -548,3 +550,61 @@ def strongly_connected_components(graph, algorithm):
...
+        if len(next_nodes) != 0:
+            for next_node in next_nodes:
+                    status = operation(next_node.name, path)
+                    if not status:
+                        continue
+                    else:
+                        break
+        else:
+            return path
+    return path
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:610]: Not necessary to track path,can be done through `operation`.

File: pydatastructs/graphs/algorithms.py [hunk 32]
```diff
@@ -548,3 +550,61 @@ def strongly_connected_components(graph, algorithm):
...
+        The node/vertex from where random walk will begin.
+    path_length:
+        The maximum number of nodes/vertices traversed by random walk.
+
+    Returns
+    =======
+
+    path: list
+        Python list of nodes after execution of random walk.
+
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:574]: Doc strings not updated.

File: pydatastructs/graphs/algorithms.py (hunk 30 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:555]: I will add `*args`, so that we can pass additional arguments if required.

File: pydatastructs/graphs/algorithms.py (hunk 32 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py]: ping @czgdp1807 
I updated them in the recent commit.

File: pydatastructs/graphs/algorithms.py [hunk 33]
```diff
@@ -637,3 +639,65 @@ def _depth_first_search_adjacency_list(
                 return None
 
 _depth_first_search_adjacency_matrix = _depth_first_search_adjacency_list
+
+def random_walk(graph, start, path_length, operation, seed=44, *args):
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:643]:
```suggestion
def random_walk(graph, start, path_length, operation, *args, **kwargs):
```

File: pydatastructs/graphs/algorithms.py [hunk 34]
```diff
@@ -637,3 +639,65 @@ def _depth_first_search_adjacency_list(
...
+    ...
+
+    >>> random_walk(g, v1, 5, f, 44, path)
+    >>> path == ['0', '1', '2', '0', '1']
+    True
+
+    '''
+
+
+    random.seed(seed)
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:686]:
```suggestion
    random.seed(kwargs.get("seed", 44))
```

File: pydatastructs/graphs/algorithms.py [hunk 35]
```diff
@@ -637,3 +639,65 @@ def _depth_first_search_adjacency_list(
...
+    path = args[0]
+    path.append(start.name)
+    op_args = args
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:689]:
```suggestion
```

File: pydatastructs/graphs/algorithms.py [hunk 36]
```diff
@@ -637,3 +639,65 @@ def _depth_first_search_adjacency_list(
...
+    start: GraphNode or it's subclasses
+        The node/vertex from where random walk will begin.
+    path_length:
+        The maximum number of nodes/vertices traversed by random walk.
+    operation: function
+        The function which is to be applied
+        on next node, in order to append it to path.
+        The prototype which is to be followed is,
+        `function_name(next_node,
+                       arg_1, arg_2, . . ., arg_n)`.
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:661]: Doesn't look good.

File: pydatastructs/graphs/algorithms.py [hunk 37]
```diff
@@ -637,3 +639,65 @@ def _depth_first_search_adjacency_list(
...
+        if len(next_nodes) != 0:
+            for next_node in next_nodes:
+                    status = operation(next_node.name, *op_args)
+                    if not status:
+                        continue
+                    else:
+                        break
+        else:
+            return None
+    return None
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:703]: Messy.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:703]: It might help in loop erased random walk. For now I will remove the if else part, since current focus is on simple random walk.

File: pydatastructs/graphs/algorithms.py (hunk 36 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:661]: Should I explain it more? Or improve the language?

File: pydatastructs/graphs/algorithms.py (hunk 37 above)
Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py]: I pushed some changes, I will resolve the errors after getting your reviews.

File: pydatastructs/graphs/algorithms.py [hunk 38]
```diff
@@ -637,3 +639,57 @@ def _depth_first_search_adjacency_list(
...
+    '''
+
+    random.seed(kwargs.get("seed", 44))
+    path = args[0]
+    path.append(start.name)
+    for i in range(path_length - 1):
+        curr_node = path[-1]
+        next_nodes = graph.neighbors(curr_node)
+        random.shuffle(next_nodes)
+        if len(next_nodes) != 0 and status:
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:692]:
```suggestion
        if next_nodes and status:
```

File: pydatastructs/graphs/algorithms.py [hunk 39]
```diff
@@ -637,3 +639,57 @@ def _depth_first_search_adjacency_list(
...
+            for next_node in next_nodes:
+                    status = operation(next_node.name, *args)
+    return None
```
Comment (czgdp1807) [inline pydatastructs/graphs/algorithms.py:695]: IMHO, this isn't correct.

Comment (Saptashrungi) [reply pydatastructs/graphs/algorithms.py:695]: The code for `random_walk` is quite simple. On integrating `operation`, its getting complicated.
Can you suggest some improvements?
Currently the code shuffles the neighboring nodes. For loop-erased random walk, if the `next_node` creates a loop, we need to consider the other neighbors of the node. Hence I have added a `for-loop`. However we do not need a loop for `simple random_walk`. 
However there is a correction, on appending `next_node`, we need to exit the inner `for-loop`.
